pkg/kubelet/kubelet_test.go

```diff
@@ -2154,6 +2156,65 @@ func TestMakeEnvironmentVariables(t *testing.T) {
 				"KUBERNETES_RO_PORT_8087_TCP_ADDR=1.2.3.7"),
 			21,
 		},
+		{
+			"simple substitution",
+			&api.Pod{ObjectMeta: api.ObjectMeta{Namespace: "unknown", Name: "test-pod"}},
+			&api.Container{
+				Env: []api.EnvVar{
+					{Name: "BAR", Value: "bar"},
+					{Name: "FOO", Value: "${BAR}"},
+				},
+			},
+			"kubernetes",
+			true,
+			util.NewStringSet(
+				"BAR=bar",
+				"FOO=bar",
+			),
+			2,
+		},
+		{
+			"slightly more complex substitution",
+			&api.Pod{ObjectMeta: api.ObjectMeta{Namespace: "unknown", Name: "test-pod"}},
+			&api.Container{
+				Env: []api.EnvVar{
+					{Name: "BAR", Value: "bar"},
```

Comment:
I agree (have been convinced) we should have a clear way to expose the downward API, and ensure substitution is consistent across container runtimes in a separate fashion.  

One problem with the json path syntax is it is an inherent mini-api.  Re: api version, we can't rely on that value for anything except pod data because the api version of the pod != the api version of other objects in the system (data from the kubelet, services, volume plugins, etc).

Maybe we should be defining an EnvVarSource on the env var (optional, invalid if Value not empty) and define a structured type for the ev source that includes how you want to expose the data:

type EnvVar struct {
  Name string
  Value string
  *EnvVarSource
}

type EnvVarSource struct {
  *PodEnvVarSource
  *HostEnvVarSource
  // ... future
}

type PodEnvVarSource struct {
  Name bool
  FieldPath string
}

type HostEnvVarSource struct {
  HostPortIP bool
}

This would be a much more describable API in general.

> On Mar 26, 2015, at 5:11 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/kubelet/kubelet_test.go:
> 
> > +				},
> > +			},
> > +			"kubernetes",
> > +			true,
> > +			util.NewStringSet(
> > +				"BAR=bar",
> > +				"FOO=bar",
> > +			),
> > +			2,
> > +		},
> > +		{
> > +			"slightly more complex substitution",
> > +			&api.Pod{ObjectMeta: api.ObjectMeta{Namespace: "unknown", Name: "test-pod"}},
> > +			&api.Container{
> > +				Env: []api.EnvVar{
> > +					{Name: "BAR", Value: "bar"},
> What will Docker's transitive expansion do for us? I looked at the code, but have forgotten already. I'd rather define a simpler mechanism to populate environment variables with Kubernetes data, and then rely upon Docker and/or the shell for more general expansion/substitution.
> 
> —
> Reply to this email directly or view it on GitHub.
>